cpp/src/arrow/compute/exec/exec_plan.cc

```diff
@@ -464,6 +464,13 @@ std::shared_ptr<RecordBatchReader> MakeGeneratorReader(
     std::shared_ptr<Schema> schema,
     std::function<Future<util::optional<ExecBatch>>()> gen, MemoryPool* pool) {
   struct Impl : RecordBatchReader {
+    ~Impl() {
+      auto st = this->Close();
+      if (!st.ok()) {
+        ARROW_LOG(WARNING) << "MakeGeneratorReader failed in finalzing reading.";
+      }
```

Comment:
Agreed, this should just be the default. I would also make the message clear that it's because Close() was not explicitly called, and include the status in the message.

Reply:
I understand, but can we include the `arrow/util/logging.h` in public APIs?

Reply:
We can move the definition to the `.cc` file, no?